exist-core/src/test/java/org/exist/backup/RestoreAppsTest.java

```diff
@@ -0,0 +1,211 @@
+
+import static org.junit.Assert.*;
+
+public class RestoreAppsTest {
+
+    @ClassRule
+    public static TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    @ClassRule
+    public static final ExistEmbeddedServer existEmbeddedServer = new ExistEmbeddedServer(true, true);
+
+    private static final String REPO_XML =
+            "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
+            "<meta xmlns=\"http://exist-db.org/xquery/repo\">\n" +
+            "    <description>Backup Test App</description>\n" +
+            "    <type>application</type>\n" +
+            "    <target>backup-test</target>" +
+            "</meta>";
+
+    /**
+     * Create an app package and generate a backup. Install a newer version
+     * of the same package and restore the backup. The newer version inside
+     * the expath repo should be preserved and not overwritten.
+     *
+     * @throws Exception in case of error
+     */
+    @Test
+    public void restoreSkipNewer() throws Exception {
+        final BrokerPool pool = existEmbeddedServer.getBrokerPool();
+        removePackage(pool);
+
+        createAndInstallApp("1.0.0");
+
+        Path backup = export(pool);
+
+        removePackage(pool);
+
+        createAndInstallApp("2.0.0");
+
+        try(final DBBroker broker = pool.get(Optional.of(pool.getSecurityManager().getSystemSubject()));
+            final Txn transaction = pool.getTransactionManager().beginTransaction()) {
+            Restore restore = new Restore();
+            TestRestoreListener listener = new TestRestoreListener();
+            restore.restore(broker, transaction, null, backup, listener, false);
+
+            assertEquals(1, listener.info.size());
+            assertTrue(listener.info.get(0).endsWith("Newer version is already installed."));
+        }
+    }
+
+    /**
+     * Create an app package and generate a backup. Install an older version
+     * of the same package and restore the backup. The newer version inside
+     * the backup should overwrite the older in the database.
+     *
+     * @throws Exception in case of error
+     */
+    @Test
+    public void restoreOverwriteOlder() throws IOException, EXistException, PackageException, PermissionDeniedException, SAXException {
+        final BrokerPool pool = existEmbeddedServer.getBrokerPool();
+        removePackage(pool);
+
+        createAndInstallApp("2.0.0");
+
+        Path backup = export(pool);
+
+        removePackage(pool);
+
+        createAndInstallApp("1.0.0");
+
+        try(final DBBroker broker = pool.get(Optional.of(pool.getSecurityManager().getSystemSubject()));
+            final Txn transaction = pool.getTransactionManager().beginTransaction()) {
+            Restore restore = new Restore();
+            TestRestoreListener listener = new TestRestoreListener();
+            restore.restore(broker, transaction, null, backup, listener, false);
+
+            assertEquals(0, listener.info.size());
+        }
+    }
+
```

Comment:
I think we need two more tests where a package version is invalid Semver 2.0, first test should be where the invalid version is already present in the db, second test should be where the invalid version is in the backup.